core/internal/runconfig/runconfig.go

```diff
@@ -165,6 +167,95 @@ func (runConfig *RunConfig) Serialize(format ConfigFormat) ([]byte, error) {
 	return nil, fmt.Errorf("config: unknown format: %v", format)
 }
 
+// Filters the configuration tree based on the given paths.
+//
+// include and exclude are lists of paths. If include is non-empty, only
+// paths that have a prefix in include are kept. If exclude is non-empty,
+// paths that have a prefix in exclude are removed. If both are empty, the
+// entire tree is returned. If both are non-empty, exlcude is applied first.
+func (runConfig *RunConfig) FilterTree(
+	include []RunConfigPath,
+	exclude []RunConfigPath,
+) RunConfigDict {
+	pathMap := dictToPathMap(runConfig.tree)
+	for _, path := range exclude {
+		prunePath(pathMap, path)
+	}
+	if len(include) > 0 {
+		var keep bool
+		for k := range pathMap {
+			keep = false
+			for _, path := range include {
+				if pathHasPrefix(*k, path) {
+					keep = true
+					break
+				}
+			}
+			if !keep {
+				delete(pathMap, k)
+			}
+		}
+	}
+	return pathMapToDict(pathMap)
+}
+
+// Checks if a given RunConfigPath has a given prefix.
+func pathHasPrefix(path RunConfigPath, prefix RunConfigPath) bool {
+	if len(path) < len(prefix) {
+		return false
+	}
+	for i, prefixPart := range prefix {
+		if path[i] != prefixPart {
+			return false
+		}
+	}
+	return true
+}
+
+// Converts of paths to values to a nested dict.
+func pathMapToDict(pathMap PathMap) RunConfigDict {
+	dict := make(RunConfigDict)
+	for path, value := range pathMap {
+		err := updateAtPath(dict, *path, value)
+		// This can only happen if update one path and then another, invalid path
```

Comment:
This doesn't seem possible now, but if someone were to use pathMapToDict throughout a run for some reason, and the user did something like update a leaf node to not be a leaf node this would trigger?

Reply:
The `pathMap` we are filtering on is a map from all leaves of the tree to their values. We use the inputs and outputs to filter out the leaves and then stitch them back together into a tree. So we can never call this with a path that goes through a leaf because all of the paths end at a leaf 🤙🏻

Reply:
I'll update the comment